s3.go

```diff
@@ -1,211 +1,313 @@
-					nextiskey = false
-					dec.Skip()
-
-					if strings.HasPrefix(kval, ".md5/") {
-						continue
-					}
-					res := dsq.Result{Entry: dsq.Entry{Key: kval}}
-					return res, true
-				}
+			index -= len(resp.Contents)
+			if index < 0 {
+				index = 0
 			}
 		}
+
+		entry := dsq.Entry{
+			Key: *resp.Contents[index].Key,
+		}
+		if !q.KeysOnly {
+			value, err := s.Get(ds.NewKey(entry.Key))
+			if err != nil {
+				return dsq.Result{Error: err}, false
+			}
+			entry.Value = value
+		}
+
+		index++
+		return dsq.Result{Entry: entry}, true
 	}
 
 	return dsq.ResultsFromIterator(q, dsq.Iterator{
 		Close: func() error {
-			return resp.Body.Close()
+			return nil
 		},
 		Next: nextValue,
 	}), nil
 }
 
+func (s *S3Bucket) Batch() (ds.Batch, error) {
+	return &s3Batch{
+		s:   s,
+		ops: make(map[string]batchOp),
+	}, nil
+}
+
 func (s *S3Bucket) Close() error {
+	close(s.limiter)
 	return nil
 }
 
+func (s *S3Bucket) s3Path(p string) string {
+	return path.Join(s.RootDirectory, p)
+}
+
+func parseError(err error) error {
+	if s3Err, ok := err.(awserr.Error); ok && s3Err.Code() == s3.ErrCodeNoSuchKey {
+		return ds.ErrNotFound
+	}
+	return err
+}
+
 type s3Batch struct {
-	s *S3Bucket
+	s   *S3Bucket
+	ops map[string]batchOp
 }
 
-func (s *S3Bucket) Batch() (ds.Batch, error) {
-	return &s3Batch{s}, nil
+type batchOp struct {
+	val    []byte
+	delete bool
 }
 
 func (b *s3Batch) Put(k ds.Key, val []byte) error {
-	return b.s.Put(k, val)
+	b.ops[k.String()] = batchOp{
+		val:    val,
+		delete: false,
+	}
+	return nil
 }
 
 func (b *s3Batch) Delete(k ds.Key) error {
-	return b.s.Delete(k)
+	b.ops[k.String()] = batchOp{
+		val:    nil,
+		delete: true,
+	}
+	return nil
 }
 
 func (b *s3Batch) Commit() error {
+	var (
+		deleteObjs []*s3.ObjectIdentifier
+		putKeys    []ds.Key
+	)
+	for k, op := range b.ops {
+		if op.delete {
+			deleteObjs = append(deleteObjs, &s3.ObjectIdentifier{
+				Key: aws.String(k),
+			})
+		} else {
+			putKeys = append(putKeys, ds.NewKey(k))
+		}
+	}
+
+	errChanSize := len(putKeys)
+	if len(deleteObjs) > 0 {
+		errChanSize++
+	}
+	errChan := make(chan error, errChanSize)
+	defer close(errChan)
+
+	for _, k := range putKeys {
+		go func(k ds.Key, op batchOp) {
+			b.s.limiter <- struct{}{}
+
+			err := b.s.Put(k, op.val)
+			errChan <- err
+
+			<-b.s.limiter
+		}(k, b.ops[k.String()])
+	}
+
+	if len(deleteObjs) > 0 {
+		go func() {
+			resp, err := b.s.S3.DeleteObjects(&s3.DeleteObjectsInput{
+				Bucket: aws.String(b.s.Bucket),
+				Delete: &s3.Delete{
+					Objects: deleteObjs,
+				},
+			})
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			var errs []string
+			for _, err := range resp.Errors {
+				errs = append(errs, err.String())
+			}
+
+			if len(errs) > 0 {
+				err = fmt.Errorf("s3ds: failed to delete objects:\n%s", strings.Join(errs, "\n"))
+			}
+			errChan <- err
+		}()
+	}
+
+	for i := 0; i < errChanSize; i++ {
+		err := <-errChan
+		if err != nil {
+			return err
```

Comment:
This will allow us to return while we're still writing. If we do that and then call close on the datastore, we'll close the limiter channel causing outstanding writes to panic.

Personally, I'd fix this by both:

1. Not using a global "limiter" channel. There's really no reason to make it global.
2. Making sure to wait for all outstanding goroutines to finish (not absolutely necessary but considered the "nice" thing to do).

Reply:
Good catch.